lxd/cluster/recover.go

```diff
@@ -0,0 +1,84 @@
+package cluster
+
+import (
+	"fmt"
+	"path/filepath"
+
+	dqlite "github.com/canonical/go-dqlite"
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/node"
+	"github.com/pkg/errors"
+)
+
+func ListDatabaseNodes(database *db.Node) ([]string, error) {
+	nodes := []db.RaftNode{}
+	err := database.Transaction(func(tx *db.NodeTx) error {
+		var err error
+		nodes, err = tx.RaftNodes()
+		return err
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to list database nodes")
+	}
+	addresses := make([]string, len(nodes))
+	for i, node := range nodes {
+		addresses[i] = node.Address
+	}
+	return addresses, nil
+}
+
+func Recover(database *db.Node) error {
+	// Figure out if we actually act as dqlite node.
+	var info *db.RaftNode
+	err := database.Transaction(func(tx *db.NodeTx) error {
+		var err error
+		info, err = node.DetermineRaftNode(tx)
+		return err
+	})
+	if err != nil {
+		return errors.Wrap(err, "Failed to determine node role.")
+	}
+
+	// If we're not a database node, return an error.
+	if info == nil {
+		return fmt.Errorf("This LXD instance has no database role.")
+	}
+
+	// If this is a standalone node not exposed to the network, return an
+	// error.
+	if info.Address == "" {
+		return fmt.Errorf("This LXD instance is not clustered.")
```

Comment:
Nit for all errors in this change, we don't end our error messages with periods.

Reply:
Fixed.